osu.Server.Spectator/Hubs/ScoreUploader.cs

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using osu.Game.Scoring;
+using osu.Server.Spectator.Database;
+using osu.Server.Spectator.Entities;
+using osu.Server.Spectator.Storage;
+using Timer = System.Timers.Timer;
+
+namespace osu.Server.Spectator.Hubs
+{
+    public class ScoreUploader : IEntityStore, IDisposable
+    {
+        /// <summary>
+        /// Amount of time (in milliseconds) between checks for pending score uploads.
+        /// </summary>
+        public double UploadInterval
+        {
+            get => timer.Interval;
+            set => timer.Interval = value;
+        }
+
+        /// <summary>
+        /// Amount of time (in milliseconds) before any individual score times out if a score ID hasn't been set.
+        /// This can happen if the user forcefully terminated the game before the API score submission request is sent, but after EndPlaySession() has been invoked.
+        /// </summary>
+        public double TimeoutInterval = 30000;
+
+        private bool shouldSaveReplays => Environment.GetEnvironmentVariable("SAVE_REPLAYS") == "1";
+
+        private readonly ConcurrentQueue<UploadItem> queue = new ConcurrentQueue<UploadItem>();
+        private readonly IDatabaseFactory databaseFactory;
+        private readonly IScoreStorage scoreStorage;
+        private readonly Timer timer;
+        private readonly CancellationTokenSource timerCancellationSource;
+        private readonly CancellationToken timerCancellationToken;
+
+        public ScoreUploader(IDatabaseFactory databaseFactory, IScoreStorage scoreStorage)
+        {
+            this.databaseFactory = databaseFactory;
+            this.scoreStorage = scoreStorage;
+
+            timerCancellationSource = new CancellationTokenSource();
+            timerCancellationToken = timerCancellationSource.Token;
+
+            timer = new Timer(5000);
+            timer.AutoReset = false;
+            timer.Elapsed += (_, _) => Task.Run(Flush);
+            timer.Start();
+        }
+
+        /// <summary>
+        /// Enqueues a new score to be uploaded.
+        /// </summary>
+        /// <param name="token">The score's token.</param>
+        /// <param name="score">The score.</param>
+        public void Enqueue(long token, Score score)
+        {
+            if (!shouldSaveReplays)
+                return;
+
+            Interlocked.Increment(ref remainingUsages);
+
+            var cancellation = new CancellationTokenSource();
+            cancellation.CancelAfter(TimeSpan.FromMilliseconds(TimeoutInterval));
+
+            queue.Enqueue(new UploadItem(token, score, cancellation));
+        }
```

Comment:
I get that this is meant to handle the case where the user hasn't completed score submission, but if the server had a large backlog (or was waiting on slow s3 operations), it feels like this could cause unexpected failures. ie. there's a chance the upload is never attempted even once, if other uploads in the queue take over 30 seconds in total.

Reply:
It is structured so that the score ID is fetched from the database first, and _only if_ it is `null` does the cancellation get considered.